aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cpp

```diff
@@ -367,6 +367,109 @@ Tensor _inverse_helper_cuda_lib(const Tensor& self) {
   return self_inv_working_copy;
 }
 
+// call cusolver gesvd function to calculate svd
+template<typename scalar_t>
+inline static void _apply_svd_lib_gesvd(const Tensor& self, Tensor& U, Tensor& S, Tensor& VT, Tensor& infos, bool compute_uv, bool some,
+  const bool calculate_all_batches,
+  const std::vector<int64_t>& batches
+) {
+  using value_t = typename c10::scalar_value_type<scalar_t>::type;
+  auto self_data = self.data_ptr<scalar_t>();
+  auto U_data = U.data_ptr<scalar_t>();
+  auto S_data = S.data_ptr<value_t>();
+  auto VT_data = VT.data_ptr<scalar_t>();
+  auto self_stride = matrixStride(self);
+  auto U_stride = !compute_uv ? 1 : matrixStride(U);
+  auto S_stride = S.size(-1);
+  auto VT_stride = !compute_uv ? 1 : matrixStride(VT);
+
+  int _batchsize = cuda_int_cast(batchCount(self), "batch size");
+  int m = cuda_int_cast(self.size(-2), "m");
+  int n = cuda_int_cast(self.size(-1), "n");
+  int lda = std::max<int>(1, m);
+  int ldvt = std::max<int>(1, n);
+
+  TORCH_INTERNAL_ASSERT(m >= n, "cusolver gesvd only supports matrix with sizes m >= n");
+
+  char jobu = compute_uv ? (some ? 'S' : 'A') : 'N';
+  char jobvt = compute_uv ? (some ? 'S' : 'A') : 'N';
+  auto handle = at::cuda::getCurrentCUDASolverDnHandle();
+
+  int lwork = -1;
+  at::cuda::solver::gesvd_buffersize<scalar_t>(handle, m, n, &lwork);
+  TORCH_INTERNAL_ASSERT(lwork >= 0, "gesvd_buffersize failed to get needed buffer size, got lwork = ", lwork);
+
+  auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
+  auto dataPtr_work = allocator.allocate(sizeof(scalar_t)*lwork);
+  auto dataPtr_rwork = allocator.allocate(sizeof(value_t)*std::min(m, n));
+
+  // VT_view shares the same memory space as VT tensor
+  at::Tensor VT_view;
+  if (compute_uv) {
+    VT_view = VT.view({-1, n, n});
+  }
+
+  // Note that gesvd returns V^H in column-major.
+  // Without a transpose here, it will be V.conj() in the actual memory space.
+  at::Tensor vt_workspace = at::empty({n, n}, VT.options()).conj();
+
+  int batchsize = _batchsize;
+  std::function<int(int)> get_batch_index = [&](int i) -> int { return i; };
+  if (!calculate_all_batches) {
+    batchsize = batches.size();
+    get_batch_index = [&](int i) -> int { return batches[i]; };
+  }
+
+  for(int _i = 0; _i < batchsize; _i++){
+    int i = get_batch_index(_i);
+
+    at::cuda::solver::gesvd<scalar_t>(
+      handle, jobu, jobvt, m, n,
+      self_data + i * self_stride,
+      lda,
+      S_data + i * S_stride,
+      U_data + i * U_stride,
+      lda,
+      vt_workspace.data_ptr<scalar_t>(), // VT_data + i * VT_stride,
```

Comment:
What is the reason why we can't write `V_data` in-place?

Reply:
added a comment explaining this

Reply:
I'd say that you just do vt = vt.conj() to activate the conj bit and return it with the conjugate bit on (and thus avoiding these copies as well). This is what we do in other linalg functions as well.

Reply:
Turning the conjugate flag on in-place would make the existing VT batches (from previously successful `gesvdj`) wrong?